File: Python/Algorithm/Is_Subsequence.py
# Complexity: O(main.size())

# For String:
# Iterative Approach:

def sseq(main, sub):
    m, n, i, j = len(main), len(sub), 0, 0
    while i < m and j < n:
        if main[i] == sub[j]: j += 1
        i += 1
    return j == n

# Recursive Approach:

def sseq(main, sub, m_sz = None, s_sz = None):
    if m_sz is None and s_sz is None:
        m_sz, s_sz = len(main), len(sub)
    if not s_sz: return True
    if not m_sz: return False
    if main[m_sz - 1] == sub[s_sz - 1]:
        return sseq(main, sub, m_sz - 1, s_sz - 1)
    return sseq(main, sub, m_sz - 1, s_sz)

# For Vector:
# Iterative Approach:

def sseq(main, sub):
    m, n, i, j = len(main), len(sub), 0, 0
    while i < m and j < n:
        if main[i] == sub[j]: j += 1
        i += 1
    return j == n

# Recursive Approach:

def sseq(main, sub, m_sz = None, s_sz = None):
    if m_sz is None and s_sz is None:
        m_sz, s_sz = len(main), len(sub)
    if not s_sz: return True
    if not m_sz: return False
    if main[m_sz - 1] == sub[s_sz - 1]:
        return sseq(main, sub, m_sz - 1, s_sz - 1)
    return sseq(main, sub, m_sz - 1, s_sz)

File: CPP/Algorithm/Is_Subsequence.cpp
// Complexity: O(main.size())
// For String:
// Iterative Approach:

bool sseq(auto &main, auto &sub) {
    int m = main.size(), n = sub.size(), i = 0, j = 0;
    while (i < m && j < n) {
        if (main[i] == sub[j]) ++j;
        ++i;
    }
    return (j == n);
}

// Recursive Approach:

bool sseq(auto &main, auto &sub, int64_t m_sz, int64_t s_sz) {
    if (!s_sz) return true;
    if (!m_sz) return false;
    if (main[m_sz - 1] == sub[s_sz - 1]) return sseq(main, sub, m_sz - 1, s_sz - 1);
    return sseq(main, sub, m_sz - 1, s_sz);
}

// For Vector:
// Iterative Approach:

bool sseq(auto &main, auto &sub) {
    int m = main.size(), n = sub.size(), i = 0, j = 0;
    while (i < m && j < n) {
        if (main[i] == sub[j]) ++j;
        ++i;
    }
    return (j == n);
}

// Recursive Approach:

auto sseq(auto &main, auto &sub, int64_t m_sz, int64_t s_sz) {
    if (!s_sz) return true;
    if (!m_sz) return false;
    if (main[m_sz - 1] == sub[s_sz - 1]) return sseq(main, sub, m_sz - 1, s_sz - 1);
    return sseq(main, sub, m_sz - 1, s_sz);
}